### ISSUE TITLE
Geometric Fabrics witj Jax

### ISSUE THREAD

**Issue by maxspahn** [cohnt/Fabrics#1]
Hi @cohnt, I recently found out about this repository and figured that you like the concept of Riemannian motion policies and fabrics. I have implemented geometric fabrics in python as well, see https://github.com/tud-amr/fabrics. In y repository, I use casadi with symbolic differentiation, whereas your code relies on jax. How was your experience with that? Is it handy/fast? I would like to hear your thought and maybe start an exchange about python implementations of fabrics. 

Best,
Max.

**Comment by cohnt** [comment]
Hello, Max!

I found the Jax was very easy to start using, but I ran into a couple little issues. For example, the log(1+exp(x)) function (which appears quite frequently in these methods) has poor numerics, and needs some special treatment (described [here](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html#numerical-stability)). It took a while to figure out that's what was causing the problems! Doing all of the symbolic computation is pretty slow -- compiling everything can take a while, and Jax hogs a lot of your GPU while it does so. Since Fabrics and co require so much tuning to get things working the way you like, this can be a pretty frustrating time sink!

I'm not actively working on RMP/Fabric stuff right now, but I am keeping it in the back of my mind. I personally prefer the Pullback Bundle Dynamical Systems way of handling velocity-dependent controllers, since placing the metric on the tangent bundle is more intuitive to me than Finsler geometry. If I was implementing something like this project again, I would probably use that instead of Fabrics.

Best,
Tommy

**Comment by maxspahn** [comment]
Hi Tommy,

Thanks for your reply. Nice to hear your experiences with jax, and also quite reassuring that is has some drawbacks as well. It is true, that the Finsler geometry can be quite unintuitive to derive a metric from. 

Thanks again for your reply. Check out our repository on fabrics, if you ever consider using it again.

Best,
Max.
